### PR TITLE
sumOn and productOn for Foldable instances

### PR DESCRIPTION
**Pull request by mel-brown**
I find myself using these often, especially for practice problems when I want to fold a structure into a number based on a some property of the value instead of the value itself. I thought `extra` would be a good place for them, since it's an automatic import for me on all my personal projects.

### CODE REVIEW

**Comment by ndmitchell** [comment]
Thanks for the PR. A couple of thoughts that make me a little hesitant:

* There is no corresponding `sumOn` etc for lists. Although there could be.
* It's not something I've ever wanted. Although there is `sumBy` in F# (and presumably OCaml) so it's not without precedence.
* The `sum` in Haskell is defined as a lazy `foldr`. That's awful, so I definitely wouldn't want to copy that mistake. But at the same time, having `sum` and `sumOn` have different behaviour isn't a great idea.

Thinking out loud, I slightly wonder if adding `sum'` to extra, being the sensible `sum`, would be a good idea. And once we have that, doing `sumOn'` starts to be more within the sensible bounds for extra. Thoughts?

**Comment by mel-brown** [comment]
Good points! I put list-specific versions of `sum'`, `product'`, `sumOn'`, and `productOn'` into `Data.List.Extra`, and renamed the corresponding generalizations in `Data.Foldable.Extra` as well.

File: src/Data/Foldable/Extra.hs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+module Data.Foldable.Extra where
```
Comment (ndmitchell) [inline src/Data/Foldable/Extra.hs:1]: Can we get an explicit export list.

File: src/Data/List/Extra.hs [hunk 2]
```diff
@@ -455,6 +457,26 @@ groupSortBy :: (a -> a -> Ordering) -> [a] -> [[a]]
 groupSortBy f = groupBy (\a b -> f a b == EQ) . sortBy f
 
 
+-- | A strict version of 'sum'.
+sum' :: (Num a) => [a] -> a
```
Comment (ndmitchell) [inline src/Data/List/Extra.hs:461]: Could we add the example `sum' [1, 2, 3] == 6` just so people have a concrete example to go from.

File: src/Data/Foldable/Extra.hs [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+
+import Data.Foldable
+
+sumOn' :: (Foldable f, Num b) => (a -> b) -> f a -> b
```
Comment (ndmitchell) [inline src/Data/Foldable/Extra.hs:5]: Worth saying `A generalisation of `Data.List.Extra.sumOn'`?

File: src/Data/Foldable/Extra.hs (hunk 1 above)
Comment (mel-brown) [reply src/Data/Foldable/Extra.hs:1]: Done!

File: src/Data/Foldable/Extra.hs (hunk 3 above)
Comment (mel-brown) [reply src/Data/Foldable/Extra.hs:5]: Completed! Also added `Data.Foldable.Extra.sum'` and `Data.Foldable.Extra.product'` for consistency.

I'm not well-versed in INLINE pragmas; I'm not sure if adding one to any of these would have any effect.

File: src/Data/List/Extra.hs (hunk 2 above)
Comment (mel-brown) [reply src/Data/List/Extra.hs:461]: Done!

File: src/Data/Foldable/Extra.hs (hunk 3 above)
Comment (ndmitchell) [reply src/Data/Foldable/Extra.hs:17]: I tend to leave INLINE pragmas off unless someone benchmarks them as having benefit - so no need to start now.

**Comment by ndmitchell** [comment]
Thanks, all merged! For info, there was a bug in productOn' I spotted when getting some tests working, which I've now fixed. I'll make a release in the next hour or so.

**Pull request merged**